src/CommunityToolkit.Maui.Core/IPopupService.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System.ComponentModel;
+
+namespace CommunityToolkit.Maui.Core;
+
+/// <summary>
+/// Provides a mechanism for displaying <see cref="CommunityToolkit.Maui.Core.IPopup"/>s based on the underlying view model.
+/// </summary>
+public interface IPopupService
+{
+	/// <summary>
+	/// Resolves and displays a <see cref="CommunityToolkit.Maui.Core.IPopup"/> and <typeparamref name="TViewModel"/> pair that was registered with <c>AddTransientPopup</c>.
+	/// </summary>
+	/// <typeparam name="TViewModel">The type of the view model registered with the <see cref="CommunityToolkit.Maui.Core.IPopup"/>.</typeparam>
+	void ShowPopup<TViewModel>() where TViewModel : INotifyPropertyChanged;
+
+	/// <summary>
+	/// Resolves and displays a <see cref="CommunityToolkit.Maui.Core.IPopup"/> and <typeparamref name="TViewModel"/> pair that was registered with <c>AddTransientPopup</c>.
+	/// The supplied <paramref name="onPresenting"/> provides a mechanism to invoke any methods on your view model in order to load or pass data to it.
+	/// </summary>
+	/// <typeparam name="TViewModel">The type of the view model registered with the <see cref="CommunityToolkit.Maui.Core.IPopup"/>.</typeparam>
+	/// <param name="onPresenting">An <see cref="Action{TViewModel}"/> that will be performed before the popup is presented.</param>
+	void ShowPopup<TViewModel>(Action<TViewModel> onPresenting) where TViewModel : INotifyPropertyChanged;
```

Comment:
I don't really like onPresenting parameter here, in my case, I would like to have a control of Presenting, Presented directly in ViewModel, I've got a custom solution for my needs, I used an interface, with proper naming it might be something like IPresentableViewModel, and inside of a service implementation you could call the method if viewModel implements this interface. There is a [link](https://github.com/prostopsih/Nikklim.App.Navigation) to my custom solution, it has more than just Popups, but anyway, I would like to extend your proposal with my ideas if there is any interest in it.

Reply:
I'm guessing your repository is private as I cannot see the contents when clicking the link.

The part I like about not using an interface is that it lets the caller use whatever implementation they like and it doesn't tie them into a constrained interface that we have to provide. It'll be good to see your code though to compare

Reply:
I apologise for it, it's public now

Reply:
I think I am inclined to leave this as-is for now. I worry about defining yet another interface for users to have to implement when it might force them down a route which won't suit their scenario. You could quite easily wrap up this method with an extension to solve your needs though:

```csharp
public void ShowPopup<TViewModel>(this IPopupService popupService, IDictionary<string, object> arguments) where TViewModel : IPresentableViewModel, INotifyPropertyChanged
{
    popupService.ShowPopup<TViewModel>(onPresenting: viewModel => viewModel.OnPresenting(arguments);
}
```

Something like the above albeit I am making a number of assumptions about your potential implementation which is probably wrong :).